api/services/content/content.proto

```diff
@@ -0,0 +1,156 @@
+}
+
+message InfoRequest {
+	string digest = 1 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+}
+
+message InfoResponse {
+	// Digest is the hash identity of the blob.
+	string digest = 1 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+
+	// Size is the total number of bytes in the blob.
+	int64 size = 2;
+
+	// CommittedAt provides the type at which the blob was committed.
+	google.protobuf.Timestamp committed_at = 3;
+}
+
+// ReadRequest defines the 
+message ReadRequest {
+	// Digest is the hash identity to read.
+	string digest = 1 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+
+	// Offset specifies the number of bytes in the blob to begin the read. If
+	// zero or less, the read will be from the start.
+	int64 offset = 2;
+
+	// size is the total size of the read. If zero, the entire blob will be
+	// returned by the service.
+	int64 size = 3;
+}
+
+// ReadResponse carries byte data for a read request.
+message ReadResponse {
+	int64 offset = 1; // offset of the returned data
+	bytes data = 2; // actual data
+}
+
+
+// WriteAction sets the action for a WriteRequst.
+enum WriteAction {
+	option (gogoproto.goproto_enum_prefix) = false;
+	option (gogoproto.enum_customname) = "WriteAction";
+	
+	// WriteActionWrite sets the action for the write request to write data.
+	//
+	// Any data included will be written at the provided offset. The
+	// transaction will be left open for further writes.
+	//
+	// This is the default.
+	WRITE = 0 [(gogoproto.enumvalue_customname)="WriteActionWrite"];
+
+	// WriteActionCommit will write any outstanding data in the message and
+	// commit the write, storing it under the digest.
+	//
+	// This can be used in a single message to send the data, verify it and
+	// commit it.
+	COMMIT = 1 [(gogoproto.enumvalue_customname)="WriteActionCommit"];
+
+	// WriteActionAbort will release any resources associated with the write
+	// and free up the ref for a completely new set of writes.
+	ABORT = -1 [(gogoproto.enumvalue_customname)="WriteActionAbort"];
+}
+
+// WriteRequest writes data to the request ref at offset. 
+message WriteRequest {
+	// Action sets the behavior of the write.
+	//
+	// When this is a write and the ref is not yet allocated, the ref will be
+	// allocated and the data will be written at offset.
+	//
+	// If the action is write and the ref is allocated, it will accept data to
+	// an offset that has not yet been written.
+	//
+	// If the action is write and there is no data, the current write status
+	// will be returned. This works differently from status because the stream
+	// holds a lock.
+	WriteAction action = 1;
+	string ref = 2;
+	int64 expected_size = 3;
+	string expected_digest = 4 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
```

Comment:
> Is it meant to be optional?

It is optional.

> My other concern would be that accepting a digest from the client leads to a risk of the blob store implementation wrongly trusting that digest in some way.

The blob store _must_ validate the content. The clients of the blob store _may_ fully trust the blob store, but the blob store should never trust the clients.

> Also, it doesn't feel right to provide the expected digest in each message in the stream, instead of a single time when setting up the stream. What should happen if the value in this field changes mid-stream?

This only needs to be provided once. It will take the last sent value before the `WriteActionCommit` is sent. If it is provided and already exists, the write will be cancelled and an error will be returned.

> Does providing an expected digest upfront enable optimizations in the blob store, such as short-circuiting the data stream if the blob already exists?

Yes, this is the main use case.